src/Bicep.Core/Utils/Result.cs

```diff
@@ -0,0 +1,51 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using Bicep.Core.Diagnostics;
+using JetBrains.Annotations;
+
+namespace Bicep.Core.Utils;
+
+public class Result<TSuccess, TError>
+    where TSuccess : class
+    where TError : class
+{
+    TSuccess? Success { get; }
+
+    TError? Error { get; }
+
+    public Result(TSuccess success)
+    {
+        Success = success;
+        Error = null;
+    }
+
+    public Result(TError error)
+    {
+        Success = null;
+        Error = error;
+    }
+
+    public bool IsSuccess()
+        => Success is not null;
+
+    public bool IsSuccess([NotNullWhen(true)] out TSuccess? success, [NotNullWhen(false)] out TError? error)
```

Comment:
Is it worth adding another overload `public bool IsSuccess([NotNullWhen(true)] out TSuccess? success)` to avoid `(out var result, out _)` when error is not needed?

Reply:
Great suggestion! Updated.